File: binary-tree/maximum-width-of-binary-tree.cpp
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if (!root) return 0;

        long long maxWidth = 0;
        queue<pair<TreeNode*, long long>> q; 
        q.push({root, 1}); // node, index

        while (!q.empty()) {
            int size = q.size();
            long long start = q.front().second;  // first index of this level
            long long end = q.back().second;     // last index of this level
            maxWidth = max(maxWidth, end - start + 1);

            while (size--) {
                auto [node, idx] = q.front();
                q.pop();

                long long newIdx = idx - start; // normalize to avoid overflow

                if (node->left)
                    q.push({node->left, newIdx * 2});
                if (node->right)
                    q.push({node->right, newIdx * 2 + 1});
            }
        }

        return maxWidth;
    }
};

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        int maxw = 0;
        if (root == null)
            return maxw;
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));
        while (!q.isEmpty()) {
            int size = q.size();
            // int minindex=q.peek().index;
            int first = 0, last = 0;
            for (int i = 0; i < size; i++) {
                Pair p = q.poll();
                TreeNode node=p.node;
                int ind=p.index;
                if(i==0)  first=p.index;
                if(i==size-1)  last=p.index;

                if(node.left!=null) {
                    q.offer(new Pair(node.left,2*ind+1));
                }
                if(node.right!=null){
                    q.offer(new Pair(node.right,2*ind+2));
                }
            }
            maxw=Math.max(maxw,last-first+1);
        }
        return maxw;
    }
}

class Pair {
    int index;
    TreeNode node;

    Pair(TreeNode node, int index) {
        this.index = index;
        this.node = node;
    }
}